### PR TITLE
Stream NARs from the remote to the destination store

### LINKED ISSUE #794: importPaths incompatibility with latest nix
I am having trouble building Hydra on nixpkgs master. Bumping hydra-unstable to 63188667b8487e931a1ab2e567c41be09bd2cc0b and building shows:

```
g++ -DHAVE_CONFIG_H -I. -I../..    -std=c++17 -I/nix/store/2xhb4hlskn33pbyph36v4wxcan56dnrw-boehm-gc-8.0.4-dev/include -I/nix/store/5rjgsqjdm71cflfb68q7m771a1rqcsk3-nix-2.4pre20200719_a79b6dd-dev/include/nix -Wall -I ../libhydra -Wno-deprecated-declarations -g -O2 -std=c++17 -include nix/config.h -c -o hydra_queue_runner-build-remote.o `test -f 'build-remote.cc' || echo './'`build-remote.cc
build-remote.cc: In member function 'void State::buildRemote(nix::ref<nix::Store>, Machine::ptr, Step::ptr, unsigned int, unsigned int, unsigned int, RemoteResult&, std::shared_ptr<State::ActiveStep>, std::function<void(StepState)>)':
build-remote.cc:447:70: error: no matching function for call to 'nix::Store::importPaths(nix::FdSource&, std::shared_ptr<nix::FSAccessor>&, nix::CheckSigsFlag)'
  447 |             destStore->importPaths(from, result.accessor, NoCheckSigs);
      |                                                                      ^
In file included from /nix/store/5rjgsqjdm71cflfb68q7m771a1rqcsk3-nix-2.4pre20200719_a79b6dd-dev/include/nix/parsed-derivations.hh:1,
                 from state.hh:13,
                 from build-remote.cc:9:
/nix/store/5rjgsqjdm71cflfb68q7m771a1rqcsk3-nix-2.4pre20200719_a79b6dd-dev/include/nix/store-api.hh:628:16: note: candidate: 'nix::StorePaths nix::Store::importPaths(nix::Source&, nix::CheckSigsFlag)'
  628 |     StorePaths importPaths(Source & source, CheckSigsFlag checkSigs = CheckSigs);
      |                ^~~~~~~~~~~
/nix/store/5rjgsqjdm71cflfb68q7m771a1rqcsk3-nix-2.4pre20200719_a79b6dd-dev/include/nix/store-api.hh:628:16: note:   candidate expects 2 arguments, 3 provided
make[3]: *** [Makefile:546: hydra_queue_runner-build-remote.o] Error 1
```

nixpkgs `hydra-unstable` depends on `nixFlakes` which now includes `Remove 'accessor' from addToStore()` https://github.com/NixOS/nix/commit/545bb2ed03001cd7a80a90f73eb500f396c043a1, but Hydra was not updated for that change.

### PR DESCRIPTION
**Pull request by edolstra**
This PR makes the processing of incoming NARs work in O(1) memory (excluding the NAR content listing). It also makes Hydra build against the current Nix master.

A few other changes:

* Removed SHA-1 hashes from the `BuildProducts` table. Since we now need to compute hashes over all regular files inside a NAR (because we might not have seen `hydra-build-products` yet) it's expensive to have a hash type that's deprecated anyway.
* Removed `TokenServer` and the `nar_buffer_size` setting.

Fixes #794.

### CODE REVIEW

File: src/lib/Hydra/Schema/BuildProducts.pm [hunk 1]
```diff
@@ -99,8 +94,6 @@ __PACKAGE__->add_columns(
   { data_type => "text", is_nullable => 0 },
   "filesize",
   { data_type => "bigint", is_nullable => 1 },
-  "sha1hash",
-  { data_type => "text", is_nullable => 1 },
```
Comment (grahamc) [inline src/lib/Hydra/Schema/BuildProducts.pm:103]: I wonder if we should keep this column and continue serving the data, but null on all future builds?

File: src/hydra-queue-runner/build-remote.cc [hunk 2]
```diff
@@ -124,11 +124,44 @@ static void copyClosureTo(std::timed_mutex & sendMutex, ref<Store> destStore,
 }
 
 
+// FIXME: use Store::topoSortPaths().
+StorePaths topoSortPaths(const std::map<StorePath, ValidPathInfo> & paths)
```
Comment (Ericson2314) [inline src/hydra-queue-runner/build-remote.cc:128]: Hah! I was just about to generalize `Store::topoSortPaths` myself!

File: src/hydra-queue-runner/nar-extractor.cc [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+using namespace nix;
+
+struct Extractor : ParseSink
+{
+    std::unordered_set<Path> filesToKeep {
+        "/nix-support/hydra-build-products",
+        "/nix-support/hydra-release-name",
+        "/nix-support/hydra-metrics",
+    };
```
Comment (grahamc) [inline src/hydra-queue-runner/nar-extractor.cc:15]: This is for the various functions that intake data from the build result and store in the db, right? Is this going to change/break other behavior?

File: src/lib/Hydra/Schema/BuildProducts.pm (hunk 1 above)
Comment (edolstra) [reply src/lib/Hydra/Schema/BuildProducts.pm:103]: I've removed the upgrade script to we don't delete any data.

File: src/hydra-queue-runner/nar-extractor.cc (hunk 3 above)
Comment (edolstra) [reply src/hydra-queue-runner/nar-extractor.cc:15]: It shouldn't change any behaviour.

File: src/hydra-queue-runner/build-remote.cc (hunk 2 above)
Comment (Ericson2314) [reply src/hydra-queue-runner/build-remote.cc]: https://github.com/NixOS/nix/pull/3864 is what I ended up using

**Comment by ajs124** [comment]
It seems like this broke `nix-support/hydra-build-products` if they reference another output.

@edolstra could you maybe have a look at #827?

**Pull request merged**